### PR TITLE
Fix for github issue #992 and additional stress tests

### PR DESCRIPTION
**Pull request by ewertons**
## Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [x] This pull-request is submitted against the `master` branch.

## Description of the changes
Fix for GitHub issue https://github.com/Azure/azure-iot-sdk-csharp/issues/992

- Passed the cancellatoionToken to _handlerLock.WaitAsync;
- Added a stress test to verify no issues incur from the change.

## Reference/Link to the issue solved with this PR (if any)
https://github.com/Azure/azure-iot-sdk-csharp/issues/992

### CODE REVIEW

File: e2e/stress/MemoryLeakTest/Program.cs [hunk 1]
```diff
@@ -127,7 +127,7 @@ public async Task<int> Method_LongRunningLeakTest_WrongEndpoint_Ok(TransportType
                 catch (TimeoutException) { }
                 catch (SocketException) { }
                 catch (TaskCanceledException) { }
-                catch (IotHubCommunicationException) { }
+                catch (IotHubCommunicationException ex2) { Console.WriteLine(ex2); }
                 finally
```
Comment (prmathur-microsoft) [inline e2e/stress/MemoryLeakTest/Program.cs]: Is it a good idea to print it to console

File: e2e/stress/RaceConditionTest/Program.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                Console.WriteLine(ex);
+                return 1;
+            }
+
+            Console.WriteLine("Test passed (" + transportType + ")");
+            return 0;
+        }
+
+        public static int Main(string[] args)
+        {
```
Comment (prmathur-microsoft) [inline e2e/stress/RaceConditionTest/Program.cs]: Why do you have a main in E2E ?

File: e2e/stress/stress.sln [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28803.352
```
Comment (prmathur-microsoft) [inline e2e/stress/stress.sln]: Why is there a version degrade?

File: e2e/stress/RaceConditionTest/Program.cs (hunk 2 above)
Comment (ewertons) [reply e2e/stress/RaceConditionTest/Program.cs]: This is 100% based on the other test already in the stress solution. I rather follow the same pattern so we don't loose focus on the purpose of testing this change.

File: e2e/stress/MemoryLeakTest/Program.cs (hunk 1 above)
Comment (ewertons) [reply e2e/stress/MemoryLeakTest/Program.cs]: This is 100% based on the other test already in the stress solution. I rather follow the same pattern so we don't loose focus on the purpose of testing this change.

File: e2e/stress/stress.sln (hunk 3 above)
Comment (ewertons) [reply e2e/stress/stress.sln]: Oh, you know, this is done automatically by Visual Studio. I'll revert this change.

File: e2e/stress/RaceConditionTest/Program.cs (hunk 2 above)
Comment (CIPop) [reply e2e/stress/RaceConditionTest/Program.cs]: @prmathur-microsoft the VSTest harness is meant for unit tests mainly. Stress testing can cause the harness itself to behave abnormally. 
The folder maybe is not ideally named: e2e\stress. This is definitely not E2E but it does exercise public API.

File: e2e/stress/RaceConditionTest/Program.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        public async Task<int> RunTest_SendTenThousandMessagesAsync(TransportType transportType)
+        {
+            const int MaxExecutionTimeMinutes = 10;
+            DeviceClient deviceClient = DeviceClient.CreateFromConnectionString(DeviceConnectionString, transportType);
+            deviceClient.OperationTimeoutInMilliseconds = 5;
+
+            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromMinutes(MaxExecutionTimeMinutes));
+
+#pragma warning disable CS0618 // Type or member is obsolete
+            deviceClient.RetryPolicy = RetryPolicyType.No_Retry;
```
Comment (CIPop) [inline e2e/stress/RaceConditionTest/Program.cs]: This isn't the right way to disable RetryPolicy. You need to use `SetRetryPolicy(IRetryPolicy)` instead.
AFAIK, this line is a no-op.

File: e2e/stress/RaceConditionTest/Program.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        private string DeviceConnectionString;
+        private const int NumberOfParallelDeviceClientCalls = 10000;
+
+        public MessageRaceConditionTest(string deviceConnectionString)
+        {
+            DeviceConnectionString = deviceConnectionString;
+        }
+
+        public async Task<int> RunTest_SendTenThousandMessagesAsync(TransportType transportType)
```
Comment (CIPop) [inline e2e/stress/RaceConditionTest/Program.cs]: Naming: Maybe better like : `Message_SendParallelMessagesAsyncNoRetry_Ok`?

File: e2e/stress/RaceConditionTest/Program.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                Console.WriteLine(ex);
+                return 1;
+            }
+
+            Console.WriteLine("Test passed (" + transportType + ")");
+            return 0;
+        }
+
+        public async Task<int> RunTest_SendThousandsOfMessagesAndCancelAsync(TransportType transportType)
```
Comment (CIPop) [inline e2e/stress/RaceConditionTest/Program.cs]: Naming: Maybe better like : `Message_SendParallelMessagesAsyncNoRetryCancelled_Timesout`?

**Comment by prmathur-microsoft** [comment]
@ewertons - Is this change still valid? if so can you please merge it in post gates passing?

**Comment by prmathur-microsoft** [comment]
Closing this PR as there has been no activity. Feel free to submit another one rebasing to master when you are ready.

**Pull request closed without merging**